src/Networking/NexusMods.Networking.Steam/Session.cs

```diff
@@ -0,0 +1,250 @@
+        _logger.LogInformation("Logged on to Steam network.");
+    }
+
+    private async Task DisconnectedCallback(SteamClient.DisconnectedCallback callback)
+    {
+        _logger.LogInformation("Disconnected from Steam network.");
+    }
+    
+    private async Task ConnectedCallback(SteamClient.ConnectedCallback callback)
+    {
+        _isConnected = true;
+        var (success, data) = await _authStorage.TryLoad();
+        if (success)
+        {
+            _logger.LogInformation("Using saved auth data to log in.");
+            var authData = AuthData.Load(data);
+            
+            _steamUser.LogOn(new SteamUser.LogOnDetails
+                {
+                    Username = authData.Username,
+                    AccessToken = authData.RefreshToken,
+                }
+            );
+        }
+        else
+        {
+            _logger.LogInformation("No saved auth data, logging in via QR code.");
+            var authSession = await _steamClient.Authentication.BeginAuthSessionViaQRAsync(new AuthSessionDetails());
+
+            authSession.ChallengeURLChanged = () => { _handler.ShowQRCode(new Uri(authSession.ChallengeURL, UriKind.Absolute), CancellationToken.None); };
+
+            _handler.ShowQRCode(new Uri(authSession.ChallengeURL, UriKind.Absolute), CancellationToken.None);
+            var pollResponse = await authSession.PollingWaitForResultAsync();
+
+            await _authStorage.SaveAsync(new AuthData
+            {
+                Username = pollResponse.AccountName,
+                RefreshToken = pollResponse.RefreshToken,
+            }.Save());
+            
+            _steamUser.LogOn(new SteamUser.LogOnDetails
+                {
+                    Username = pollResponse.AccountName,
+                    AccessToken = pollResponse.RefreshToken,
+                }
+            );
+        }
+    }
+
+    /// <summary>
+    /// Wraps an async action in a synchronous action.
+    /// </summary>
+    private Action<T> WrapAsync<T>(Func<T, Task> action)
+    {
+        return arg => Task.Run(async () => await action(arg));
+    }
+    
+    public async Task<ProductInfo> GetProductInfoAsync(AppId appId, CancellationToken cancellationToken = default)
+    {
+        await ConnectedAsync(cancellationToken);
+        var jobs = await _steamApps.PICSGetProductInfo(new SteamApps.PICSRequest(appId.Value), null);
+        if (jobs.Failed)
+            throw new Exception("Failed to get product info for app " + appId.Value);
+
+        return ProductInfoParser.Parse(jobs.Results![0]);
+    }
+
+    /// <summary>
+    /// Performs a login if required, returns once the login is complete
+    /// </summary>
+    private async Task ConnectedAsync(CancellationToken cancellationToken)
+    {
+        if (!_isConnected)
+        {
+            _steamClient.Connect();
+        }
+        
+        while (!_isLoggedOn)
+        {
+            await _callbacks.RunWaitCallbackAsync(cancellationToken);
+        }
```

Comment:
Since many methods just pass `default(CancellationToken)` to this method, we should wrap the token and make sure we have a timeout in place.

```suggestion
        using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        cts.CancelAfter(delay: TimeSpan.FromMinutes(5));

        while (!_isLoggedOn)
        {
            await _callbacks.RunWaitCallbackAsync(cts.Token);
        }
```